### PR TITLE
Add simplified param.depends and watch mechanisms

### PR DESCRIPTION
**Pull request by philippjfr**
Supersedes https://github.com/ioam/param/pull/250 by making a number of simplifications:

* No ``viewable`` argument for ``param.depends``
* No ``cache`` argument for ``param.depends`` (can be added back later)
* Renamed ``eager`` keyword ``watch`` in ``param.depends``

### CODE REVIEW

File: param/parameterized.py [hunk 1]
```diff
@@ -192,6 +193,66 @@ def __get__(self, obj, type_=None):
...
 
 
+def _getattrr(obj, attr, *args):
+    def _getattr(obj, attr):
+        return getattr(obj, attr, *args)
+    return reduce(_getattr, [obj] + attr.split('.'))
+
+
+# (thought I was going to have a few decorators following this pattern)
+def accept_arguments(f):
```
Comment (jbednar) [inline param/parameterized.py]: Would be helpful to have docstrings on some of these new functions

File: param/parameterized.py [hunk 2]
```diff
@@ -424,6 +487,35 @@ def _set_instantiate(self,instantiate):
...
+
+        super(Parameter, self).__setattr__(name, value)
+
+        if old is not NotImplemented:
+            for subscriber in self.subscribers[name]:
+                subscriber(Change(what=name,attribute=self._attrib_name,obj=None,cls=self._owner,old=old,new=value))
+
+    # TODO: this is a python 3.6+ thing; we can presumably do this in
+    # Parameterized.__new__ for older pythons. (And merge with objtype
+    # slot.)
```
Comment (jbednar) [inline param/parameterized.py]: @ceball, can you work out what to do here for when not in 3.6?

File: param/parameterized.py [hunk 3]
```diff
@@ -424,6 +487,35 @@ def _set_instantiate(self,instantiate):
...
+    #
+    # (not specific to this change, but regarding attrib_name, owner:
+    # what if someone re-uses a parameter object across different
+    # classes? maybe we should raise an error if attrib name,owner
+    # already set)
```
Comment (jbednar) [inline param/parameterized.py]: Seems like that would be polite of us but is not required for merging.

File: param/parameterized.py [hunk 4]
```diff
@@ -472,24 +564,38 @@ def __set__(self,obj,val):
         object stored in a constant or read-only Parameter (e.g. the
         left bound of a BoundingBox).
         """
+        # TODO: simplify this method!
```
Comment (jbednar) [inline param/parameterized.py]: Or document it, at least.

File: param/parameterized.py [hunk 5]
```diff
@@ -979,6 +1085,65 @@ def inspect_value(self_,name): # pylint: disable-msg=E0213
 
         return value
 
+
+    def params_depended_on(self_,name):
```
Comment (jbednar) [inline param/parameterized.py]: This and several more below need docstrings...

File: param/parameterized.py [hunk 6]
```diff
@@ -1553,6 +1718,19 @@ def __init__(self,**params):
...
         object_count += 1
 
+        # TODO: should move to param namespace? (like _param_value
+        # etc)
+        self._param_subscribers = {}
+
+        # add watched dependencies
+        #
+        # TODO: This isn't great. Also note anything here will happen
+        # for every instantiation.
```
Comment (philippjfr) [inline param/parameterized.py]: This bit seems to be pretty expensive, it seems to take about 4x longer than before to instantiate a parameterized class.

**Comment by ceball** [comment]
With #255, does this cover what yuo need for now?

When merged, I'll make an issues about things that must be cleaned up before release, and things that could be improved/added in the future.

**Comment by philippjfr** [comment]
>With #255, does this cover what yuo need for now?

Yes I think so. Looking forward to seeing this merged.

**Comment by ceball** [comment]
Who's handling that - me or you? I don't mind either way, just let me know if I should do it...

**Comment by philippjfr** [comment]
>Who's handling that - me or you? I don't mind either way, just let me know if I should do it...

Missed your question about subscribe on the other PR. Happy to merge both once that's deleted.

File: param/parameterized.py (hunk 6 above)
Comment (ceball) [reply param/parameterized.py]: Did it get any better? I'd expect it to be a bit better than before (and minimal impact when there's no watching), but there are still improvements I plan to make (future PR).

File: param/parameterized.py (hunk 5 above)
Comment (jlstevens) [reply param/parameterized.py]: I'm also not sure about the names yet but we can iterate on that outside this PR.

Comment (ceball) [reply param/parameterized.py]: I agree - will be part of an issue covering things that need to be sorted out before the next release.

**Comment by jlstevens** [comment]
Looking at the diff I do have a number of suggestions to make. To avoid slowing things down, I won't write anything down just yet so that we can merge sooner rather than late.

**Comment by ceball** [comment]
> Looking at the diff I do have a number of suggestions to make

I have some suggestions too. But yes, I think until we've settled on the functionality we want, we should hold off. Unless you're talking about something that's an error...

**Comment by philippjfr** [comment]
I just double checked and after updating ``panel`` and my holoviews PR everything still works great.

**Pull request merged**